src/corelibs/U2View/src/ov_msa/mca_reads/McaAlternativeMutationsWidget.cpp

```diff
@@ -0,0 +1,169 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#include "McaAlternativeMutationsWidget.h"
+
+#include "../McaEditorSequenceArea.h"
+
+#include <U2Core/McaDbiUtils.h>
+#include <U2Core/MsaDbiUtils.h>
+#include <U2Core/MultipleChromatogramAlignmentObject.h>
+#include <U2Core/U2AttributeUtils.h>
+#include <U2Core/U2Mod.h>
+#include <U2Core/U2SafePoints.h>
+#include <U2Core/U2OpStatusUtils.h>
+
+namespace U2 {
+
+const QString McaAlternativeMutationsWidget::ALTERNATIVE_MUTATIONS_CHECKED = "ALTERNATIVE_MUTATIONS_CHECKED";
+const QString McaAlternativeMutationsWidget::ALTERNATIVE_MUTATIONS_THRESHOLD = "ALTERNATIVE_MUTATIONS_THRESHOLD";
+
+McaAlternativeMutationsWidget::McaAlternativeMutationsWidget(QWidget* parent)
+    : QWidget(parent) {
+    setupUi(this);
+
+}
+
+void McaAlternativeMutationsWidget::init(MultipleAlignmentObject* _maObject, MaEditorSequenceArea* _seqArea) {
+    SAFE_POINT(_seqArea != nullptr, "MaConsensusModeWidget can not be initialized: MaEditorSequenceArea is nullptr", );
+    SAFE_POINT(_maObject != nullptr, "MaConsensusModeWidget can not be initialized: MultipleAlignmentObject is nullptr", );
+
+    seqArea = qobject_cast<McaEditorSequenceArea*>(_seqArea);
+    SAFE_POINT(seqArea != nullptr, "MaConsensusModeWidget can not be initialized: McaEditorSequenceArea is nullptr", );
+
+    mcaObject = qobject_cast<MultipleChromatogramAlignmentObject*>(_maObject);
+    SAFE_POINT(mcaObject != nullptr, "MaConsensusModeWidget can not be initialized: MultipleChromatogramAlignmentObject is nullptr", );
+
+    mutationsGroupBox->setChecked(false);
+    mutationsThresholdSlider->setValue(99);
+
+    updateValuesFromDb();
+
+    connect(mutationsGroupBox, SIGNAL(toggled(bool)), this, SLOT(sl_updateAlternativeMutations()));
+    connect(mutationsThresholdSlider, SIGNAL(valueChanged(int)), this, SLOT(sl_updateDb()));
+    connect(mutationsThresholdSpinBox, SIGNAL(valueChanged(int)), this, SLOT(sl_updateDb()));
+    connect(updateMutationsPushButton, SIGNAL(pressed()), this, SLOT(sl_updateAlternativeMutations()));
+}
+
+void McaAlternativeMutationsWidget::sl_updateAlternativeMutations() {
+    U2OpStatus2Log os;
+    U2UseCommonUserModStep userModStep(mcaObject->getEntityRef(), os);
+    Q_UNUSED(userModStep);
+
+    mcaObject->updateAlternativeMutations(mutationsGroupBox->isChecked(), mutationsThresholdSlider->value(), os);
+    CHECK_OP(os, );
+
+    updateDb(os);
+    CHECK_OP(os, );
+}
+
+void McaAlternativeMutationsWidget::sl_updateDb() {
+    U2OpStatus2Log os;
+    updateDb(os);
```

Comment:
Why is there a `U2UseCommonUserModStep` variable in the `sl_updateAlternativeMutations`, but not here?

Reply:
Because it won't work for this change. The "undo/redo" framework works with `U2ObjectDbi`, but changes about "Alternative mutations" settings are written to `U2AttributeDbi`. I don't think that this behavior should be changed

Reply:
When I click Update, the Undo button enables. Does this mean that changes have been made to Object (not Attribute)?